var/spack/repos/builtin/packages/py-hpc-beeflow/package.py

```diff
@@ -34,22 +34,22 @@ class PyHpcBeeflow(PythonPackage):
     #Python dependencies
     depends_on("py-flask@2.0.0:", type=("run"))
     depends_on("py-fastapi@0.109.2", type=("run")) #This version doesn't exist yet in Spack
-    depends_on("py-uvicorn@0.27.1", type("run")) #This version doesn't exist yet in Spack
-    depends_on("py-pyyaml@6.0.1:", type("run"))
-    depends_on("py-flask-restful@0.3.9", type("run"))
-    depends_on("py-cwl-utils@0.16:", type("run"))
-    depends_on("py-apscheduler@3.6.3:", type("run")) #This version doesn't exist yet in Spack
-    depends_on("py-jsonpickle@2.2.0:", type("run"))
-    depends_on("py-requests@:2.28.0", type("run"))
-    depends_on("py-requests-unixsocket@0.3.0:", type("run"))
-    depends_on("py-python-daemon@2.3.1:", type("run"))
-    depends_on("py-gunicorn@20.1.0:23.0.0", type("run"))
-    depends_on("py-typer@0.5.0:", type("run"))
-    depends_on("py-cffi@1.15.1:", type("run"))
-    depends_on("py-celery+redis+sqlalchemy@5.3.4:", type("run"))
-    depends_on("py-docutils@0.18.1:", type("build", "run"))
-    depends_on("py-graphviz@0.20.3:", type("run"))
-    depends_on("graphviz", type("run"))
-    depends_on("py-networkx@3.1", type("run"))
+    depends_on("py-uvicorn@0.27.1", type=("run")) #This version doesn't exist yet in Spack
+    depends_on("py-pyyaml@6.0.1:", type=("run"))
+    depends_on("py-flask-restful@0.3.9", type=("run"))
+    depends_on("py-cwl-utils@0.16:", type=("run"))
+    depends_on("py-apscheduler@3.6.3:", type=("run")) #This version doesn't exist yet in Spack
```

Comment:
@aquan9, two more hints:
- It would be good if you count run `spack install py-hpc-beeflow` on some machine before you push commits.
- If the type is only "run", use `type="run"`.

As these versions do not exist yet in `spack`, I suggest the following:

The these versions are really easy to add to spack, you can add these versions in this PR.

If they are not easy to add, open new PRs that add these versions (one PR for each package).

Note: You are expected to check the `pyproject.toml` files (or other sources) for the version dependenies of those new versions and add those version dependencies as well to these packages.

Reply:
Gotcha. I'll go through the pyproject.toml for fastapi@0.109.2 and validate that those dependencies are still correct.

Reply:
Woah so this sort of sets off a cascade of things that I need to update and validate right? Because beeflow depends on fastapi, but the version isn't there for that, so I need to add it, then validate the dependencies from the pyproject.toml file for fastapi, and I see it needs starlette, but the version of starlette I need doesn't exist, so it's like a continuous rinse and repeat. 

Is the Spack project working on anything to try to automate some of this?

Reply:
I guess it wasn't so bad, it only took 30 minutes, but it's a process that seems fraught with the potential for user error. 

What I mean is I don't really have a way to validate that the new dependency-dependency packages I've added work correctly besides the use of the dependency in the package that I'm adding.

Reply:
Haven't looked at what you've done yet but am glad it wasn't too bad.  Automating it can be problematic too given the amount of variability across different types of projects, repositories, etc.

Reply:
Yes, there is some work on discovering new versions and updating them on a regular basis,
but to let this fully flourish, more dedicated contributors are likely needed, I think.

I've asked this same question, at least on about if for Python packages, the `pyproject.toml` can be automatically checked for new dependencies of new versions, but the now former Python maintainer (who did an unimaginable amount of work on manually checking dependencies on reviews) said that it can't be done automatically for now as many possible sources for versions (maybe older setup.py/setup.cfg/requriements.txt) should be checked as well.

However, I think we should start somewhere, and that still needs some brainpower to automate getting the e.g. now more and more often used `pyproject.toml` format and extract those versions.

There are some tools that do that, but those that do exist only maintain the latest version of a package, but `spack` needs to support multiple versions of packages, so because of this special capability that spack needs to provides, I have not found an "off-the shelf" solution that could be just used.

The tooling would have to be able to add version dependencies with "when=@<version>:", but sometimes new versions also remove dependencies, so translating all of that will be quite some work.

Of course, baby steps would already help to have checking of missing dependencies for newly submitted additions of newer versions. But such minimal check of missing dependencies would be a good addition to my tool that I write and already use to build (and thereby) check pull requests: https://github.com/spack/build-quality-tools

It could keep the extracted sources and after it built a package, it gets the build location already (using the build log file that is in the same directory) and it could get the pyproject.toml file afte each build and check if it contains packages or versions that are not mentioned in the updated recipe.

Contributions 💯 % welcome! :)

Reply:
I see... Thank you for the very detailed reply. I'll take a look at that tool you've put together.